File: leetcode/path-sum-iii/tree-and-recursion.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func helper(root *TreeNode, sum int, started bool) int {
	if nil == root {
		return 0
	}

	ans := 0
	if sum == root.Val {
		ans += 1
	}

	if false == started {
		ans += (helper(root.Left, sum, started) + helper(root.Right, sum, started))
	}

	ans += (helper(root.Left, sum - root.Val, true) + helper(root.Right, sum - root.Val, true))

	return ans
}

func pathSum(root *TreeNode, sum int) int {
	if nil == root {
		return 0
	}

	return helper(root, sum, false)
}

File: leetcode/maximum-depth-of-n-ary-tree/tree-and-recursion.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
public:
    int maxDepth(Node *root, int depth) {
        if (NULL == root) { return depth; }

        int ans = depth + 1;
        for (auto & child : root->children) {
            ans = max(ans, maxDepth(child, depth + 1));
        }
        return ans;
    }

    int maxDepth(Node* root) {
        int ans = 0;
        if (NULL == root) { return ans; }

        return maxDepth(root, ans);
    }
};